jax/api.py

```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
+    return out, 0
+  batching.primitive_batchers[fun_p] = fun_batch
+
+  staged_fun_p = core.Primitive('staged_' + name)
+  def fun_partial_eval(trace, *tracers, **params):
+    tracers = tuple(map(trace.instantiate_const, tracers))
+    avals = [t.aval for t in tracers]
+    pvals_in = [pe.PartialVal((a, core.unit)) for a in avals]
+    jaxpr, pval_out, consts = pe.trace_to_jaxpr(lu.wrap_init(fun_impl, params),
+                                                pvals_in, instantiate=True)
+    consts = trace.new_instantiated_const(core.pack(consts))
+    eqn = pe.JaxprEqn((consts,) + tracers, None, staged_fun_p, (), False, False,
+                      dict(params, jaxpr=jaxpr))
+    return pe.JaxprTracer(trace, pval_out, eqn)
+  pe.custom_partial_eval_rules[fun_p] = fun_partial_eval
+
+  def staged_fun_translation(c, xla_consts, *xla_args, **params):
+    consts_shapes = tuple(c.GetShape(xla_consts).tuple_shapes())
+    xla_consts = tuple(xla.xla_destructure(c, xla_consts))
+    arg_shapes = map(c.GetShape, xla_args)
+    built_c = xla.jaxpr_computation(params['jaxpr'], (), consts_shapes, *arg_shapes)
+    return c.Call(built_c, xla_consts + xla_args)
+  xla.translations[staged_fun_p] = staged_fun_translation
+
+  return CustomTransformsFunction(fun, fun_p)
+
+def _check_custom_transforms_type(name, fun):
+  if type(fun) is not CustomTransformsFunction:
+    msg = ("{} requires a custom_transforms function as its first argument, "
+          "but got type {}.")
+    raise TypeError(msg.format(name, type(fun)))
+
+def defjvp_all(fun, custom_jvp):
+  """Define a custom JVP rule for a ``custom_transforms`` function.
+
+  If ``fun`` represents a function with signature ``a -> b``, then
+  ``custom_jvp`` represents a function with signature ``a -> T a -> (b, Tb)``,
+  where we use ``T x`` to represent a tangent type for the type ``x``.
+
+  In more detail, ``custom_jvp`` must two arguments, both tuples of length equal
+  to the number of positional arguments to ``fun``. The first argument to
+  ``custom_jvp`` represents the input primal values, and the second represents
+  the input tangent values. ``custom_jvp`` must return a pair where the first
+  element represents the output primal value and the second element represents
+  the output tangent value.
+
+  Defining a custom JVP rule also affects the default VJP rule, which is derived
+  from the JVP rule automatically via transposition.
 
-  @wraps(fun)
-  def traceable(*args, **kwargs):
-    # TODO(mattjj): pytrees to jaxtupletrees
-    return fun_p.bind(*args, **kwargs)
-  traceable.primitive = fun_p
+  Args:
+    fun: a custom_transforms function.
+    custom_jvp: a Python callable specifying the JVP rule, taking two tuples as
+      arguments specifying the input primal values and tangent values,
+      respectively. The tuple elements can be arrays, scalars, or (nested)
+      standard Python containers (tuple/list/dict) thereof. Must be functionally
+      pure.
```

Comment:
Can the output be a pytree? Also briefly mention what the output is here.

Reply:
Done.